Initial support for microchip cec1702

fabled opened the pull request:
Since "Initial support for microchip cec1702 #5383" got closed and had lot of cruft. I decided to make new PR. This is essentially the same set of changes rebased on top of current master. One commit was merged earlier, and it's removed. The other commits were updated to reflect changes in current master.

fabled committed:
arch: arm: Add Microchip CEC1702 SoC

Basic support for the SoC and the UARTs in it. UART divider is
reset to register value zero which the chip treats as 3. Thus the
default baud rate for the chip is 38400.

The SoC has external interrupt controller (EC) which is configured
to pass-through all interrupts to NVIC. Support for the EC may need
to be implemented better if/when deep sleep modes are implemented.

A post-build script cec1702-image.py will create SPI flashable image.

Signed-off-by: Timo Teräs <timo.teras@iki.fi>

fabled committed:
boards: Add support for SecureIoT1702

Add basic secureiot1072 support with serial ports. Tested with
hello_world to run and work as expected.

Signed-off-by: Timo Teräs <timo.teras@iki.fi>

fabled committed:
serial/uart_ns16550: support cec1702 alternate clock

CEC1702 ns16550 compatible UART has alternate high-speed
clock selectable with the high bit of the baud rate divisor
register.

Signed-off-by: Timo Teräs <timo.teras@iki.fi>

dbkinder commented on boards/arm/secureiot1702/doc/secureiot1702.rst:
```diff
@@ -0,0 +1,82 @@
+- OTP programmable in CEC1702
+- Two expansion headers compatible with MikroElektronika mikroBUS™ Expansion interface
+
+More information can be found from `SecureIoT1702 website`_ and
```
```suggestion
More information can be found on the`SecureIoT1702 website`_ and
```

dbkinder commented on boards/arm/secureiot1702/doc/secureiot1702.rst:
```diff
@@ -0,0 +1,82 @@
+- Two expansion headers compatible with MikroElektronika mikroBUS™ Expansion interface
+
+More information can be found from `SecureIoT1702 website`_ and
+`CEC1702 website`_. The SoC programming information is available
```
```suggestion
`CEC1702 website`_, and SoC programming information is available
```

dbkinder commented on boards/arm/secureiot1702/doc/secureiot1702.rst:
```diff
@@ -0,0 +1,82 @@
+
+More information can be found from `SecureIoT1702 website`_ and
+`CEC1702 website`_. The SoC programming information is available
+in `CEC1702 datasheet`_.
```
```suggestion
in the `CEC1702 datasheet`_.
```

dbkinder commented on boards/arm/secureiot1702/doc/secureiot1702.rst:
```diff
@@ -0,0 +1,82 @@
+Please refer to the `SecureIoT1702 schematics`_ for the pin routings.
+Additional devices can be connected via mikroBUS expansion interface.
+
+Programming and Debugging
```
Add a Building section that includes an example of building one of the sample apps, as done in most other board documents.

dbkinder replied:
@fabled Will you be adding a building section to the document?

fabled replied:
@dbkinder Yes, I'll take most of what is in `mec2016evb_assy6797`and/or `mec15xxevb_assy6853` since they are essentially the same CPU. Though, perhaps I try to get my python image generator script integrated also to avoid dependency on the Microchip Windows tool. I am busy for this and next week. But will try to get this polished before end of this month - perhaps even this week if I get a good timeslot.

dbkinder replied:
Please re-request my review when the building section is here.

fabled replied:
Added small section here based on four other similar docs. The building is also already described in more detail a bit below.

dbkinder commented on boards/arm/secureiot1702/doc/secureiot1702.rst:
```diff
@@ -0,0 +1,82 @@
+Flashing
+========
+
+The produced zephyr.bin needs to be converted to bootable image using
```
```suggestion
The produced ``zephyr.bin`` must be converted to a bootable image using
```

dbkinder commented on boards/arm/secureiot1702/doc/secureiot1702.rst:
```diff
@@ -0,0 +1,82 @@
+Debugging
+=========
+
+The MCU has standard JTAG debugging features. But the author of this port
```
Delete "But the author of this port has not used it."  Can you provide a reference link about this in the microchip documentation you're linking to?

fabled replied:
@franciscomunoz Seems you added standard Debugging example for the MEC boards. Does that work? If yes, it should work also for CEC and I could just add the same here.

franciscomunoz replied:
@fabled  It works on the MEC1701 and MEC1501. It could work on the CEC1702 for sure, but very basic features as it may require extra support from the tool vendor.

fabled replied:
Ok. Maybe I copy that over and state is as "should" work. I can also look into cloning the parts of the Building steps. But the `objcopy` step is not needed for me. I also have python based tool that can do the `zephyr.bin` conversion to SPI flashable image, but I think that is worth a separate commit. @franciscomunoz Do you know if MEC is using same SPI image format as CEC? I think it does, it's mostly the same thing. I also saw you got the gpio and pinmux stuff done, those could be reused in CEC side too.

franciscomunoz replied:
@fabled It worked on the 1701, You just need to update the base address. Another thing I remember is the script ran with python2.7. It never ran with python3.X for me. Right now there are gpio,pimux, i2c, eSPI(subsystem is being reviewed). More stuff is being cooked. Drivers should work on the CEC (in fact this is why there is an Xec). However, a new HAL is going to be needed for the CEC to abstract the registers in the same way it is done for the MEC1501.

fabled replied:
> @fabled However, a new HAL is going to be needed for the CEC to abstract the registers in the same way it is done for the MEC1501.
I am thinking it might be better to try to avoid Microchip HAL. Similar how the UART code is shared, and use DT for providing all the base addresses, interrupts etc. Perhaps the gpio,pinmux,etc. could be made to have the register offsets internally, and take everything else from DT and be independent from the HAL. What do you think?
Also does MEC have the same QMSPI? If yes, should that be renamed to xec_qmspi and also used in the MEC code?

franciscomunoz replied:
@fabled I can't make any of this decisions as there are several people involved. The easiest path for would be to have the CEC1702 HAL.

fabled commented:
Any ideas why the `Shippable` tests are failing? They seem to fail at building untouched SPI drivers with errors such as:
```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_ll_stm32.c: In function 'spi_stm32_get_err':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_ll_stm32.c:54:13: error: implicit declaration of function 'LL_SPI_ReadReg'; did you mean 'LL_GPIO_ReadReg'? [-Werror=implicit-function-declaration]
...
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/spi/spi_ll_stm32.c:461:12: error: 'spi_stm32_init' defined but not used [-Werror=unused-function]
```

fabled commented:
Building/Flashing and Debugging sections updated per request. Added Python script to do the SPI image creation, so there is now no dependency on proprietary Microchip SDK (Windows) tools. Rebased on top of master, and retested `hello_world` on the SecureIoT1704 board.

Any ideas why `Documentation` or `Shippable` tests are failing? See https://github.com/zephyrproject-rtos/zephyr/pull/16669#issuecomment-500003094 for notes about the `Shippable` test.

fabled commented:
Any ideas why these come?
```
zephyr/drivers/spi/spi_cec_qmspi.c:276:13: warning: Macro is deprecated
  .spi = (QMSPI_INST_Type *) DT_MICROCHIP_CEC_QMSPI_0_BASE_ADDRESS,
             ^~~~~~~~~~~~~~~~~~~~~
zephyr/drivers/spi/spi_cec_qmspi.c:288:13: warning: Macro is deprecated
       &api_funcs);
             ^~~~~~              
```

fabled commented:
Anything more to fix? The test suite is now fixed with the exception of License one due to new HAL, and documentation should be fixed per request. Any more comments or concerns?
@galak Ping.

galak commented:
> zephyr/drivers/spi/spi_cec_qmspi.c:276:13: warning: Macro is deprecated
>   .spi = (QMSPI_INST_Type *) DT_MICROCHIP_CEC_QMSPI_0_BASE_ADDRESS,
>              ^~~~~~~~~~~~~~~~~~~~~
> zephyr/drivers/spi/spi_cec_qmspi.c:288:13: warning: Macro is deprecated
>        &api_funcs);
>              ^~~~~~

what test or sample are you seeing this on?

fabled commented:
> > zephyr/drivers/spi/spi_cec_qmspi.c:276:13: warning: Macro is deprecated
> > .spi = (QMSPI_INST_Type *) DT_MICROCHIP_CEC_QMSPI_0_BASE_ADDRESS,
> > ^~~~~~~~~~~~~~~~~~~~~
> > zephyr/drivers/spi/spi_cec_qmspi.c:288:13: warning: Macro is deprecated
> > &api_funcs);
> > ^~~~~~
> 
> what test or sample are you seeing this on?

Not with any test/sample. It comes only when compiling in the driver. So it happens only on my customized config.

Similar to what the current in-tree spi nor driver says when it's compiled:
```
zephyr/drivers/flash/spi_nor.c:276:13: warning: Macro is deprecated
  } else if ((DT_JEDEC_SPI_NOR_0_ERASE_BLOCK_SIZE == SZ_64K)
```

galak commented:
> > > zephyr/drivers/spi/spi_cec_qmspi.c:276:13: warning: Macro is deprecated
> > > .spi = (QMSPI_INST_Type *) DT_MICROCHIP_CEC_QMSPI_0_BASE_ADDRESS,
> > > ^~~~~~~~~~~~~~~~~~~~~
> > > zephyr/drivers/spi/spi_cec_qmspi.c:288:13: warning: Macro is deprecated
> > > &api_funcs);
> > > ^~~~~~
> > 
> > 
> > what test or sample are you seeing this on?
> 
> Not with any test/sample. It comes only when compiling in the driver. So it happens only on my customized config.
> 
> Similar to what the current in-tree spi nor driver says when it's compiled:
> 
> ```
> zephyr/drivers/flash/spi_nor.c:276:13: warning: Macro is deprecated
>   } else if ((DT_JEDEC_SPI_NOR_0_ERASE_BLOCK_SIZE == SZ_64K)
> ```

give me a way to reproduce and I can provide more guidance.

galak commented on drivers/serial/uart_ns16550.c:
```diff
@@ -293,6 +293,11 @@ static void set_baud_rate(struct device *dev, u32_t baud_rate)
 		 */
 		divisor = ((dev_cfg->sys_clk_freq + (baud_rate << 3))
 					/ baud_rate) >> 4;
+#if defined(CONFIG_SOC_SERIES_CEC)
```
This needs to be more abstracted.  We should introduce an additional DTS compatible for the uart on the SoC Series and use the `DT_COMPAT_<FOO>` define to base this ifdef on.  Also we need to define the magic number

fabled replied:
Updated per request. I appended the additional compatible string to the list so the normal config options will work as-is. Invented some symbolic names (based on CEC data sheet) for the flag, and used the system clock config as proper. Is something more needed?

galak reviewed (changes requested):
I'd recommend pulling out the SPI driver changes into a separate PR so we can get the base SoC support merged.

fabled commented:
> I'd recommend pulling out the SPI driver changes into a separate PR so we can get the base SoC support merged.

Dropped the SPI driver commit. I will submit it after this is merged. Also filed earlier #17902 to address some of the API issues for quad support.

fabled commented:
@galak Ping

galak commented on dts/arm/microchip/cec1702.dtsi:
```diff
@@ -0,0 +1,53 @@
+			reg = <0x400f2400 0x400>;
+			interrupts = <40 0>;
+			clock-frequency = <1843200>;
+			current-speed = <38400>;
```
Any reason you have the default here as 38400?  Typically we just leave it out.

galak reviewed (changes requested):
Need to update the HAL bits now that we have a separate module for the microchip HAL

galak commented on boards/arm/secureiot1702/Kconfig.defconfig:
```diff
@@ -0,0 +1,22 @@
+config UART_NS16550_PORT_0
+	def_bool y
+
+config UART_NS16550_PORT_1
```
In the dts you are only enabling PORT_0, so why enable PORT_1 here?

galak commented on soc/arm/microchip_cec/Kconfig.soc:
```diff
@@ -0,0 +1,15 @@
+	select CPU_CORTEX_M
+	select CPU_CORTEX_M4
+	select CPU_CORTEX_M_HAS_SYSTICK
+	select CPU_HAS_FPU
```
Does it have an FPU, If so than the dts needs updating.  Should be `compatible = "arm,cortex-m4f"`

galak reviewed (commented):
A few minor additional comments.

fabled commented:
Seems to be failing due to not able to checkout correct revision of hal_microchip module. Perhaps https://github.com/zephyrproject-rtos/hal_microchip/pull/4 could be merged first, and then trigger rebuild here.

fabled commented:
FYI @scottwcpg

fabled commented:
@galak Ping.

@franciscomunoz @scottwcpg @albertofloyd @MaureenHelm @ioannisg 
Also I have a more fundamental question for Zephyr maintainers. Most of the existing MEC/CEC work that came from Microchip is based on "split" HAL headers for MEC1501. But seems those are new and experimental and not available for any other chip from the family making the "xec" drivers completely bound to the MEC1501.

I would like to use Zephyr on the CEC1702 which is of same family. And when I asked Microchip if they could publish the same style headers, they started asking financial justification for the work. So it seems they are not yet sure if they want to ship these headers for all chips.

Would it be more portable to use the original SDK HAL headers for the *xec* drivers so they actually work through out the CEC/MEC family?

albertofloyd commented:
@fabled let me provide some context on this and some ideas that could help.

Early on when mec1701/mec1501 support started, HAL-based drivers were the plan all along. After been working for a while on Zephyr a few enhancements in the HAL have been discussed, hence the style of headers you see under mec1501 folder. 

In the long term plan these are some of the changes considered.
1) Allow to get device tree base address for SoC blocks instead of been hardcoded in HAL headers.  You can see that in progress in @scottwcpg pending pull request. 
We have worked in the past with MEC14xx and know that most of SoC blocks are almost same except base addresses.  Also note this change in HAL is not that disruptive to existing drivers.

2) Replace MEC1701 header with same header style than mec1501 reusing as much as possible the existing definitions for other mec1501 and adjust the base address in the device tree.

I think in your case you could start with mec1501 headers as your CEC17 HAL and adjust them based on CEC17 datasheet, as I mentioned before you will find a lot of commonalities.

fabled commented:
> @fabled let me provide some context on this and some ideas that could help.
> 
> Early on when mec1701/mec1501 support started, HAL-based drivers were the plan all along. After been working for a while on Zephyr a few enhancements in the HAL have been discussed, hence the style of headers you see under mec1501 folder.

Ok. However, I see the new style HAL being both improvement and a step backwards in some aspects. Perhaps it's still work-in-progress, and the minor inconvenience can be fixed?

It is very good thing that it's components based, and thus can be much more easier reused in compatible chips. That also means it does not clutter #define namespace etc.

1. However, the bitfield structure definitions are removed. Maybe this was intentional? Though, it means the code becomes now much more less readable when all the `MCHP_xxx` variables need to be used in manual bitfield twiddling.
2. The components headers still expose the SoC specific addresses. e.g. `components/qmspi.h` gives `MCHP_QMSPI_BASE_ADDR`, the GIRQ stuff, and even the system clock. If these would be moved to the SoC header in this case `MEC1501hsz.h`, the `components/qmspi.h` could be shared between SoCs. In fact, since the addresses are in DTS, we would need *only* these in the peripheral drivers without any SoC specific header (that would probably needed still in SoC init).
3. Some of the macroes still access hardware directly e.g. `MCHP_GIRQ_BLK_SETEN` from `component/ecia.h` which is used in the GPIO driver. Perhaps this issue with GIRQ side of things only? And seems it's being worked with the GIRQ driver basics in PR #18684.

> In the long term plan these are some of the changes considered.
> 
>     1. Allow to get device tree base address for SoC blocks instead of been hardcoded in HAL headers.  You can see that in progress in @scottwcpg pending pull request.
>        We have worked in the past with MEC14xx and know that most of SoC blocks are almost same except base addresses.  Also note this change in HAL is not that disruptive to existing drivers.

Is it? I was using the original CEC HAL and as the block structures are defined as `struct` with name,  I could use them in similar manner as in the new code. Please see, e.g. the QMSPI driver I wrote in https://github.com/fabled/zephyr/commit/8870821cbc89c18cd366141bf59ee99e5d945310. 

>     2. Replace MEC1701 header with same header style than mec1501 reusing as much as possible the existing definitions for other mec1501 and adjust the base address in the device tree.

@scottwcpg Are the new HAL headers already generated automatically, or are they manual work?

Would it be possible to strip the base addresses from current component headers and make them generic and shareable? E.g. move *all* the base addresses and GIRQ numbers etc. to the SoC header, and keep only the register defines and structs in the component header. Perhaps doing some exceptions for the GIRQ side of things until the real driver gets worked out.

If the headers are manual work, and the above plan sounds acceptable, I would be happy to help with the work and make the current tree work with CEC1702, at least with the peripherals I am currently needing.

> I think in your case you could start with mec1501 headers as your CEC17 HAL and adjust them based on CEC17 datasheet, as I mentioned before you will find a lot of commonalities.

Yes. That's why am calling in for the discussion how to share the existing drivers, and HAL :) Though, there are some peripherals unique to CEC also. I believe one example is the GP-SPI, to which I just wrote a basic driver. See https://github.com/fabled/zephyr/commit/6767cbd95e340e3768bcf29ebc8808d9072b33c3 (MEC has eSPI instead).

fabled commented:
Rebased on top of master. Direction on how to do HAL would be appreciated.
